### ISSUE TITLE
PP and Context Parallel composability issue -- forward hooks

### ISSUE THREAD

**Issue by fegin** [pytorch/torchtitan#1728]
The current CP API requires users to wrap the entire forward and backward passes within a `context_parallel()` context manager. This design has proven unpopular, and there is a consensus to remove it. As we consider replacing the TorchFunctionMode/monkey patching approaches with forward hooks, we are also exploring the possibility of using forward hooks to replace the `context_parallel()` context manager. More details can be found in [[pytorch/pytorch#163185](https://github.com/pytorch/pytorch/pull/163185)].

Currently, PP requires users to provide an SPMD parallelize function that applies all SPMD parallelisms to each pipeline stage. This approach works because SPMD parallelisms typically target inner modules or submodules of the model, offering finer granularity than a stage. When an SPMD parallelism installs forward hooks (e.g., TP), the installation proceeds as expected.

However, CP differs from other SPMD parallelisms in that it may need to register forward hooks on the entire model to transform inputs (such as BlockMask) for all subsequent submodules.

## Proposed Solution

One proposal is to expose certain pipeline parallelism information to the SPMD `parallel_fn`. Currently, the function signature is:

```python
parallel_fn(module, parallel_dims, job_config)
```

We could extend this to:

```python
parallel_fn(module, parallel_dims, job_config, pp_is_first_stage, pp_is_last_stage, pp_micro_batch_size)
```

This explicit signature is for illustration purposes; the actual implementation could use a dataclass or another structure to encapsulate this information.

cc., @H-Huang @tianyu-l @wwwjn @XilunWu

**Comment by tianyu-l** [comment]
Can we just install hooks to `nn.Embedding` to shard inputs? Non-first PP ranks don't have embedding layer so shouldn't be affected.
https://github.com/pytorch/torchtitan/blob/main/torchtitan/models/llama3/model/model.py#L336

**Comment by fegin** [comment]
I prefer not. I would rather change the parallelize_fn to annotate if this stage is the first or not . 

Reasons: 1) I still prefer passing in BlockMask. Then this hook cannot rewrite BlockMask. 2) Even if we call init_mask_fn inside the forward, then we will not be able to do so if we install the hook to the nn.Embedding. If we install the hook to nn.Embedding, We cannot call init_mask_fn before nn.Embedding because the inputs are sharded yet. We cannot call the init_mask_fn after nn.Embedding because we lost the eos_id after nn.Embedding.

Since CP requires modifying the inputs, I think it makes sense to let CP understand some minimal information of PP.

**Comment by tianyu-l** [comment]
There are multiple challenges to solve:
1. First PP stage should shard the inputs; later PP stage shouldn't.
2. When PP does microbatching, if it sees block mask as one of inputs, the block mask should be updated at microbatch level, for correctness.
3. Ideally each microbatch should have its own load balancing decision, for efficiency.

**For 2**
There are two things we can pass around: `eos_id` or block mask.
- With `eos_id` PP microbatching logic is simpler.
- With block mask, CP load balancing is made possible.
> If we install the hook to `nn.Embedding`, We cannot call `init_mask_fn` before `nn.Embedding` because the inputs are sharded yet.

I didn't understand this part. Wouldn't CP inputs sharding requires mask already be generated, for load balancing purpose?

**For 3**

Without considering load balancing:
Since CP inputs sharding is a one-time thing before each iteration, I think it's OK to put it in data post-processing explicitly, instead of registering a pre-forward hook. This way we can just use the [`pp_has_first_stage`](https://github.com/pytorch/torchtitan/blob/main/torchtitan/train.py#L63) flag in train.py to branch. This way we can avoid PP pollute SPMD parallelization.

I do think we should do load balancing for each microbatch:
It means CP inputs sharding should happen AFTER microbatch split in PP. Even for this we can do "only register hook in the `pp_has_first_stage` branch in train.py, but yeah maybe it's better to pass `pp_has_first_stage` to parallize fn instead of writing ad hoc CP code in train.py.

**Comment by fegin** [comment]
>> If we install the hook to nn.Embedding, We cannot call init_mask_fn before nn.Embedding because the inputs are sharded yet.

> I didn't understand this part. Wouldn't CP inputs sharding requires mask already be generated, for load balancing purpose?

So this mostly concerns how are we going to do load balancing? Since nn.Embedding is not a module that accepts mask, how are we going to shard the input in the pre-forward hook nn.Embedding if we are going to do load balancing? Unless we are going to wrap nn.Embedding again, lol.

> Since CP inputs sharding is a one-time thing before each iteration, I think it's OK to put it in data post-processing explicitly, instead of registering a pre-forward hook. This way we can just use the [pp_has_first_stage](https://github.com/pytorch/torchtitan/blob/main/torchtitan/train.py#L63) flag in train.py to branch. This way we can avoid PP pollute SPMD parallelization.

Well then there is not much difference between the context_parallel() context manager and the data post-processing - just some API needed to be called before forward(). I can make `context_parallel()` not a context manager since it doesn't have to be a context manager any more.

Then it's just an API needs to be called (we can change context_parallel to something else) for non-pp case and pp_has_first_stage case. 

> I do think we should do load balancing for each microbatch:

@XilunWu already does this when generating the load balancing masks, looking into each batch (batch size = 1).

My opinion
1. Installing the hook to nn.Embedding is not a good idea. It is too ad hoc and I don't know how to do load balancing.
2. Either having a explicit data post-processing or passing PP information to parallelize_fn work for me.

**Comment by tianyu-l** [comment]
> Either having a explicit data post-processing or passing PP information to parallelize_fn work for me.

I'm interpreting "explicit data post-processing" as "CP does inputs sharding before PP microbatch split". If so, how would you be able to do load balancing for each microbatch?

To me the only way is to install hooks to do load balancing. This means after PP microbatch, when CP does inputs sharding, what it sees is microbatch.

Then the only question is:
- do we generate block masks before PP microbatch split, so that PP has to understand masks. See https://github.com/pytorch/torchtitan/issues/1723#issuecomment-3315057399
- or do we delay block masks after PP microbatch split, so that PP microbatching logic stays simple. This means we do the attention mask init **in the model**, rather than sending masks to the model as inputs (instead we send `eos_id`).

**Comment by fegin** [comment]
> If so, how would you be able to do load balancing for each microbatch?

Assuming local batch size is 8, let's say we have 8 samples. We will generate a load balancing indices tensor with the first dimension is 8 to represent how each sample is reshuffled. And we will actually reshuffle each sample individually based its load balancing index. The resulting batch is then sharded along the sequence dimension.

So each sample (batch size 1) is load balanced. The finest granularity of PP microbatching is also batch size 1, one sample. So we can still perform load balancing even if we perform CP sharding data before PP microbatch split.

cc., @XilunWu

**Comment by tianyu-l** [comment]
Oh it's good to know that load balancing is done at sample level anyways. Yeah if so CP sharding can indeed happen before PP microbatching.

Then I would prefer "having a explicit data post-processing", compared with "passing PP information to parallelize_fn"

**Comment by fegin** [comment]
Just realized that this doesn't work because freq_cis doesn't have batch dimension -- it is broadcasted before being multiplied with the embedding.

**Comment by fegin** [comment]
We added positions to the input. We also added the feature to partition BlockMask in PP. We can close this issue.
